src/devices/src/virtio/iovec.rs

```diff
@@ -0,0 +1,604 @@
+    /// the given offset.
+    ///
+    /// # Returns
+    ///
+    /// The number of bytes read (if any)
+    pub fn read_at(&self, buf: &mut [u8], offset: usize) -> Option<usize> {
+        self.sub_region(offset, buf.len()).map(|sub_region| {
+            let mut bytes = 0;
+            let mut buf_ptr = buf.as_mut_ptr();
+
+            sub_region.into_iter().for_each(|iov| {
+                let src = iov.iov_base.cast::<u8>();
+                // SAFETY:
+                // The call to `copy_nonoverlapping` is safe because:
+                // 1. `iov` is a an iovec describing a segment inside `Self`. `IoVecSubregion` has
+                //    performed all necessary bound checks.
+                // 2. `buf_ptr` is a pointer inside the memory of `buf`
+                // 3. Both pointers point to `u8` elements, so they're always aligned.
+                // 4. The memory regions these pointers point to are not overlapping. `src` points
+                //    to guest physical memory and `buf_ptr` to Firecracker-owned memory.
+                //
+                // `buf_ptr.add()` is safe because `IoVecSubregion` gives us `iovec` structs that
+                // their size adds up to `buf.len()`.
+                unsafe {
+                    std::ptr::copy_nonoverlapping(src, buf_ptr, iov.iov_len);
+                    buf_ptr = buf_ptr.add(iov.iov_len);
+                }
+                bytes += iov.iov_len;
+            });
+
+            bytes
+        })
+    }
+}
+
+/// This is essentially a wrapper of a `Vec<libc::iovec>` which can be passed to `libc::readv`.
+///
+/// It describes a write-only buffer passed to us by the guest that is scattered across multiple
+/// memory regions. Additionally, this wrapper provides methods that allow reading arbitrary ranges
+/// of data from that buffer.
+#[derive(Debug)]
+pub(crate) struct IoVecBufferMut {
+    // container of the memory regions included in this IO vector
+    vecs: Vec<iovec>,
+    // Total length of the IoVecBufferMut
+    len: usize,
+}
+
+impl IoVecBufferMut {
+    /// Create an `IoVecBufferMut` from a `DescriptorChain`
+    pub fn from_descriptor_chain(mem: &GuestMemoryMmap, head: DescriptorChain) -> Result<Self> {
+        let mut vecs = vec![];
+        let mut len = 0usize;
+
+        for desc in head {
+            if !desc.is_write_only() {
+                return Err(Error::ReadOnlyDescriptor);
+            }
+
+            // We use get_slice instead of `get_host_address` here in order to have the whole
+            // range of the descriptor chain checked, i.e. [addr, addr + len) is a valid memory
+            // region in the GuestMemoryMmap.
+            let iov_base = mem
+                .get_slice(desc.addr, desc.len as usize)?
+                .as_ptr()
+                .cast::<c_void>();
+            vecs.push(iovec {
+                iov_base,
+                iov_len: desc.len as size_t,
+            });
+            len += desc.len as usize;
```

Comment:
Mhh, seems #3666 got lost in the rebase :(

Reply:
Damn! Let me find the commit and push it.

Reply:
Done!